### PR TITLE
Added LoggerMessage.Define overloads to disable IsEnabled check

### LINKED ISSUE #45290: Add LoggerMessage.Define overload to disable IsEnabled check
## Background and Motivation

`LoggerMessage.Define` allows skipping various overheads in logging (parsing the template etc) and returns delegate that performs logging and includes a check if that logging category is enabled before performing the call to log. In some cases it's important that check be moved higher up the stack to avoid transformation of parameters before calling log, for e.g.

```C#
public class Connection
{
    private string _connectionId;
    public string ConnectionId => _connectionId ??= GenerateConnectionId();
}

public static class Log
{
    private static readonly Action<ILogger, string, Exception> _connectionStart =
        LoggerMessage.Define<string>(
            LogLevel.Debug,
            new EventId(1, "ConnectionStarted"),
            "Connection {ConnectionId} started");

    public static void ConnectionStarted(ILogger logger, Connection connection)
    {
          _connectionStart(logger, connection.ConnectionId, null);
    }
}
```

The above logic ends up eagerly allocating the `ConnectionId` even when the log level isn't enabled. The code could be changed to this:

```diff
public static void ConnectionStarted(ILogger logger, Connection connection)
{
+    if (logger.IsEnabled(LogLevel.Debug))
+    {
        _connectionStart(logger, connection.ConnectionId, null);
+    }
}
```

It would end up checking if the category is enabled multiple times when the category is on.

## Proposed API

```diff
namespace Microsoft.Extensions.Logging
{
    public static class LoggerMessage
    {
+        public static Action<ILogger, T1...T6, Exception> Define<T1...T6>(LogLevel logLevel, EventId eventId, string formatString, bool skipEnabledCheck);
    }
```

The ... are overloads T1 to T6 (or more).

## Usage Examples

```C#
public class Connection
{
    private string _connectionId;
    public string ConnectionId => _connectionId ??= GenerateConnectionId();
}

public static class Log
{
    private static readonly Action<ILogger, string, Exception> _connectionStart =
        LoggerMessage.Define<string>(
            LogLevel.Debug,
            new EventId(1, "ConnectionStarted"),
            "Connection {ConnectionId} started",
            skipEnabledCheck: true);

    public static void ConnectionStarted(ILogger logger, Connection connection)
    {
        if (logger.IsEnabled(LogLevel.Debug))
        {
            _connectionStart(logger, connection.ConnectionId, null);
        }
    }
}
```

## Risks

None.

### PR DESCRIPTION
**Pull request by gfoidl**
Fixes https://github.com/dotnet/runtime/issues/45290

/cc: @davidfowl @KalleOlaviNiemitalo

### CODE REVIEW

**Review by viv718918 — APPROVED** [review]

File: src/libraries/Microsoft.Extensions.Logging.Abstractions/src/LoggerMessage.cs [hunk 1]
```diff
@@ -128,14 +128,35 @@ public static class LoggerMessage
...
+            {
+                return (logger, exception) => Log(logger, exception);
+            }
+
             return (logger, exception) =>
             {
                 if (logger.IsEnabled(logLevel))
                 {
-                    logger.Log(logLevel, eventId, new LogValues(formatter), exception, LogValues.Callback);
+                    Log(logger, exception);
```
Comment (davidfowl) [inline src/libraries/Microsoft.Extensions.Logging.Abstractions/src/LoggerMessage.cs:159]: I'm assuming the codegen differences are negligible?

Comment (gfoidl) [reply src/libraries/Microsoft.Extensions.Logging.Abstractions/src/LoggerMessage.cs:159]: The methods in the closure / display-class look like
```c#
internal void <Define>b__1(ILogger logger, Exception? exception)
{
    <Define>g__Log|0(logger, exception);
}

internal void <Define>b__2(ILogger logger, Exception? exception)
{
    if (logger.IsEnabled(logLevel))
    {
        <Define>g__Log|0(logger, exception);
    }
}

internal void <Define>g__Log|0(ILogger logger, Exception? exception)
{
    logger.Log(logLevel, eventId, new LogValues(formatter), exception, LogValues.Callback);
}
```
so the JIT should inline `<Define>g__Log|0` and in effect the codegen be the same.

Before it was
```c#
internal void <Define>b__0(ILogger logger, Exception? exception)
{
    if (logger.IsEnabled(logLevel))
    {
        logger.Log(logLevel, eventId, new LogValues(formatter), exception, LogValues.Callback);
    }
}
```

Comment (davidfowl) [reply src/libraries/Microsoft.Extensions.Logging.Abstractions/src/LoggerMessage.cs:159]: I figured it should inline. I was gonna check but https://sharplab.io/ is down 😢

**Review by davidfowl — APPROVED** [review]
LGTM

File: src/libraries/Microsoft.Extensions.Logging.Abstractions/src/LoggerMessage.cs (hunk 1 above)
Comment (gfoidl) [reply src/libraries/Microsoft.Extensions.Logging.Abstractions/src/LoggerMessage.cs:159]: Wanted to do the same, unfortunately...we're too dependent on sharplab 😄

Comment (davidfowl) [reply src/libraries/Microsoft.Extensions.Logging.Abstractions/src/LoggerMessage.cs:159]: We all need to sponsor @ashmind

Comment (ashmind) [reply src/libraries/Microsoft.Extensions.Logging.Abstractions/src/LoggerMessage.cs:159]: Restarted, should work now — thanks for letting me know!

Comment (gfoidl) [reply src/libraries/Microsoft.Extensions.Logging.Abstractions/src/LoggerMessage.cs:159]: It doesn't inline, but it's a tail-call ([sharplab](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIAYACY8gOgGEIAbTmMDASwgA7XAG4a9JuRTjajZiwDinCMGyd+AL2wDhsyQoAqMBBlkSAzFKRNSDADIQA5k5hQAsjFy5srmgG8aBmCmK2YbZlIAHgBJRxc3NAYAUQQwGAAHXSEAfgA+BgARGAAzfiEYAAp4+xgANxhOBhUnWobOJOSGoQwYgBMGepge/qTmRhLofB0AZQwocqck4AguBlwAa34M5KFsYB4+tgALXg2ASiCQwOoQu4dnADV1AFcvADEpnQw3BkmoaYYH5QBgAXgYbFgOhg8WenDeuE+AO+bkq/0BcwWQiWgwQGV4Pz6ADlsPgYH0AArYKCkmDAjgvHogBh0c7mW73YLEFAPJyVOLOVwglpCzppTLZHK49JZQRCS4czk3TmckVuFjxSotNqNJJDEZ9JIVADuvLhCLRXyBbnOevFsuESVhry87HUnDUYAushVDAAvjQrpz+CUGJVNttdvtDiczgrfcrfSFiAB2MNqqB2mXZc5ggqajNZiVytlB+4B6hlu6p9OCxLS4vCXOgvJV65t4OhrV1qAsGK4KMHcnd1pDTjneNJ4KJqd3As9osO+U+2cV31+lfBCtljILOrQ6xml2Iq3AiFQn7O+EfU+o8Z/K2YxZJcoYXH4vjkklkynU2n0iBGQwSdpw7Bh92FJ5jyRQEz3BE0j2vE9kWtKBLRQp9sVLStFXuV8GGwPgXnUBkejBZooKQmCUV7c0YG/V1SLMMCQzDQiMGIzgmIYABCcETA/QkGN/GkyQAoCQPuGckwwY4oAgU0EIAQSgJwXjJHpUmzOVKmw9dA1w6s0xaOjkNgtxN39Ay7lfNw9iaKE+mETgAE91nmF4+EQhEGGZGIACUYGwPoAHkhFc+x+FwDAogAaRgFy6KpfgoCicYklUAArAl8lbXDpOrMJyAiABWKIrwRMVtOEHIxnIOgCjYd1PQ2dlfV3fh9x+BhHOctyKpvFCzwAfXRGi2pVYgrAG3BqkohFqNQh8hptMCCpVUbbxBcExtQyy7m3QzkyseLEpdZLUvShgspygpZKigBtfDyj6EwAF01rAu5XAwL72yOqdWMqF6TDBcFWT+pVIaTGsENOpLsBStL6oy4Bsr4fJKgAIn8EKFiccp1EWv0saSTaVt7PH+AJ+zFr02cDuhu5ZPkxSYFNGIhFehAQpeDAQpKfzsGxGAtMbIRKj2MkIBKYGuZMCd9pVNdlesyarHw7iWwYcgJs5KaGBiXZ1LcHRoDihKEaRq6boxvICkUOljbJGkMGgXTPoB+5mHIJg03u3AHroN6la3NX9asCAGigBZXqkRhDAgTC+WbApyfM3tFt0vX7iNoQTdd6ApAsJQnfzl2zbQyS7nW/W00djBndNt2q9DqzcIrQ6DdsqASkImBDfiIUAjLAB6ceGCiXB1OmKAXLyu5x9HhgAHUFh+XACIopxBh6OeWDHieolH6f8Fn+fD+XqIMj/fAGClmBQSx7UxyxvJdnmNzjX4bgGGAAfjTrx+EIa6ICA7NDHCwY+N9RILxCEvSeMDST31pE/fUvQ+hv36NdUMskB7oKgaPJB+A4HBAQdfW+KCyRP2itCN+hhTi70/gwN2f8AFAOGCwCEws2EEU4LgCABEQG2wwIQ4hpCGDkOIVQx+WMTDVSEPQxh8jxY9UaNCAYrDwHDE/mI2+EipGUIfk/XawI37EBKtkFhgiwAXgHtgSeuAYAD1saUJ+4wsajwKGSbwvgB4yxYYwiholYChmMVjWhPxPEFGFgMYJpJQkyLQfabI0S9GwMvpPDALl8TSPCYYOYdC8gMIHtk/EODAn+LRgSaxvDAH8GtEIQhZTMj6LLNyXkUQCkYGhHkOao52jbx1B0FI3QMGDDGaMBg3SDyRJgFVcWUoVFLjGGVGZPwFlLlqvHAopibQ53Iafc+BiJ6xi9JvVieCGAE26Ig2+iTwkv3aNEhgUVd7RnJAfXC5CT4z2pBfb5R88moOfs4YZb9hm1P/gwMApwvSfOgW0wFV9Uy4DyFEMAeQPIwGPpi15oZhgfL6IQ1FEiVhrH7IOQ4/TIVPMaNhTJU8/lzxOcvAAQjAGmm8HEtH4GAdQ118SuzlOsSA+IvmLyPr8s+/zWV3NEkkiJPSonFMYfwV6PQQz8F+P8SpoqIDisRRk5FWScmtIVfkwpKqSksLNRUq5cyoWcvKPq8p/xmlmvEYy0lSkQFT2cTC0JT8YiFCihkCAPghwvNkjoXefRN5XN5fypoBrm4itwGK/xIC+hhojTAElKY0VlhDbmyNPAGAcppjMTNXSrUwD6esgeczsJdysMMdSvJhkjxNUymVLLMnxATccWNkAeiIzAYw/AEa3yvR6T/ckDAfE+FcLgLhJSnGLq8Murwi7sBuVHXOkBTiRANM6vYjIGQND8qsX0HQ2AJXwInuugeS6/HctgAwHNZaF3ADcq9PuLxOBvliesY4gFOADAqNHXhhKhwDBdQ43cEA+ieSscMOoKVhAaVEYy6VxyyyGBpOkcidA0Dh0kUfI5sqB3OCHbG6kA8XhOIGLqnu7Ez2vKEA0aK1MdAipQ1iHer12gGuw1w6ZpwN0tE3rgMDgGBgdVnj/fdwhD2ZPKLtEVTHSiAeWp+mAwAXguEWEIgYw6Gj30ESobEcBgR333NeB9ZCpVUf7bhYohmd7gnIGRnCkqr6uYBf53kdG3zzEIhsPVrgKg0iaCUFQpoAlXOwBeq9fHhBrskwPaToHwNmewBZ6zThbNuHsy6JzFGr54eo7hTmmnhDkVIL5xlgW5WDsCbG441NjgaCcMcYDID9hCC+CmkEjIBIEgXegzjeqUuXr5elkB8WFLLD5p+wRw23wQDwVAb+G7+XacqZkubaW0MIF4HzEVrDooGoqz81rZYV7UiECZ8EFhms9oeya9rMa3xdb671/rDBjSnAnS4l4McdF/ASxUkwF2rFvJuxehdKHSmCIcX3H+EP80Sa8E2uTEHOM5uvfYzJmP4QfpdVcsAEPYBkXYxh7JRpttCIIqlhb2Q4DfzjuT7Hd2XPMqCyEZIMci7gnQORw5gu2u0Y6zOrwtj+DQp4YyWAkBo4fLZ/N69Iqi64BctFGAd9bF4GOBlEEDjr14HmAarrYA/iIwp6U4db5YAAEcXgpS8Ops+5J+AHhRJqjLuGvt3EhKe5N5ESofeC72/DJqiQs+0yxou9SBDYm3pu3xK6uEzA/Fqndv2t4ihMyTpw0A3Kyby5Zt8af7FCDcq+nPIfpdlkTxUciSBfOtrUcFPq7koCodGTo/o3bY+h8fcvGI+BtcNNcoGoKG9WddBHyx+Sd8rk3OGI4gNriShP1fGklvfahfOavsC6hWN+BKIHv6lxQar89GjTY4Q0cwvo5ASvg06TSQSINuEV5GfK9BpQVZuIuL/DBOWN8fgSSWuP2e+dmYfA0YGemMOPzSfRxVvE1M5DYC5XBY0QRXAfPMoBdO/BfffLGCA/oF5coWhIQdITeczZ3JtWkcCcrXHNyBjQYD3AVS5U4NyJgvVHwMkQ2YlY/ePWPC/WRHgEoDAG/P4FKaKHfe/NxSgyZTBLxH/EhRlKQp+fGfreQpxUdOJJxFQigqgjQvILQuVNKQtdFUw5oYWJwAgqAPoJ+bFLGSRAoPgqpdGDAd9fBHgzgAtItXCf/YqP+VYFNIVSuMGSoCwiBWQzodQnqbrYCMsOAmsGQ0RZIII2afQ4CSyQ6cfLA4LHAvAlhAg9YYg7VEw3fB/CwmgkQHpegndQQnNEoEoNwKHBzBEfnKrCfM/eVZBR5UoOQ1VAeMoKAJQsgvfJ+RozQo1X/HQoxEFAoww3gYQOosw+Y9QtJJY7QntUlMgzgJwlwtwrGDwrw/FPVERAImvbgziEIv/IqGwclaIsAkEPieIlI7I5I1fVIvrdI/KMCGsHibIlgXIzifItI1A9uFrUojAzmU9dQLQHdHhWglo4jJLIJBwuYtQ1fF5aKQfPgfoyearNzSQ1Yy/dVeQ8uNwPlV5DVAQEgkEXVbRMZawlYi1EFB+Ok1gnEt5AhA4l4pAyA56Q0eOKUB+cifObgWAsCbBcEdVNuBiWU2kIoyXSjREoYhuIdAeekhYe3dVHRAvNkouDknRMk4+QY0ITjN8bBfwa5OkUQeEz7HUyrBgPUvVGUwUzeAhcQmrQqKUhgNUp0n6V04ojAqIcoU4deJye3LxdpSOaOWOAee8ROZOD2EEr2GsNUnIKUfoFgTM+YRYbOMsKMoYuPIMpErmDnHdEHOkOMvVGnSHMiERV5f0oI2pYWbbZsjsnElgkQlpNdKo9Be4mAbsnwgQgrZg9YVg9Va0ik0/T0+JO+cJPstwN+X1a6apLyVhSAGfLg7+WSQJN5ERLko4uwk4s46AC4q4rxG46nWnKHDsoU7srRRhTckEYhOkCya6HbPbeZZQxw7Ec4kxdQJxTwxY0eUlZMyItYKEyCn4gE78hU3MtMJU8Eb8vsPoTU9Aqs2MhkjABMrw+CqONwNMhCpoJC/hSoERKULKdC30IGLKTs++QDccT2BmGsPufhGANuSs72NMNit5BI9BbBAAMkkpSDyMqAkr6DhKEvISIvjIgETNFIopjhNPtK9LpAAAlTcOBXpsya5QTMK8KKyaBDoe57J1ggpDhbBvJBoM4x8QgOouoXFhAlD7wiROK6JyIsZJZOLzgsZLIPKDxepwp91h0oAHo3oGBFpigNB8AGk3BN5wQnSAByNAHKhgLKkALK/0cKvcSKoKJyaK+OBgdOHQEqzqMqvvSqyKaKZGBqaq3o+iWkDKhA00ZqmKcYPpBlQFUMkKQwZIPyXBLrblbwI3A4NyIYEEQzH+eNPTCoDASAKEC9JIK5f+GHSAOOKveTGFYdEWaxEAQ+BgOACtevTbDRAiTeB6GYLYDIRwZNXAZEjABK1hbADpTQNwCALKkzB1HpL0d0dSrKgYWFakao4WC6q6xadql0BiVdbiHNNXDAefV8dHGGoQcjA2GaRaYESoe8XaZilUIGXaMGDi+U7i2cFmBSHqhgFSNSbDPy7gMWJcSWWkGWdCQERWMCISu4KmGmImK0ciXaHOO4CCcCXAYAWUxA5ONlT3CDVEOEmyMiDNYWTmbmEjNufCYYPobW0GHaK0DUYYJwWSSWkIEHedMMTWoQI2hASeONR26uf6BmfCVNIQNlIjGAR28id4F6H2/uR23mnQJILK/wPK+2x2vULmV2tuGyLsGO+WJ20EcGBgaS0A7232/29Ol21Ot2qGL2X0CeRPBgMDHgTeSYRkMQkujaSmk2lCROyaFMFutA6GfCMAFQJxYO9IPOxKoO3O1OsOjACOv0PKr2vuv21OuOw2wuq2zsMMbuvNaevO/ieOwu6GOA30OoWWlgJSZHLmUeg+3APPYWcMflB22egunWq6lO7mfm+uzkB+422+kwdu9uBmRoJxbepmR9RKsWtKu+fXMdJ25kJ0kGBAB6NAVEpwIQbDN6B6XyXaZOBKv0CrD2siXaZK/gVK4EAewOze7mAWX1FyUepIJKxoPBtK6YjKfEHOkOm+le3u4ex+xepMPe4AA+o+voE+pSM+m+CWV+hAeh4YNe1Oy60VLWyRgAah1ifoZilv3sPoYb4eGg6uRvYEAh6GLKTlLOxF0jVtnA0aRq6oPr6D4d2hYBmBeGAHGHooYYkZ1vkZ82WkBFwfwbcH9quqnrYdBiuvIEUaUeCC4Z4bUf4cEYvpweoa8agFjphR7pgGcYCfcZ0E8dof9tcafv/uCBEfIhYeSf8adtcc/pVk5EFpCBqrfHBDCZLME3LI7jxqKkYGFsJk4ARvDJdIYAiu6icTMC/uOgcCin6vqgKDomRrzERuvGRpzl6fTIiPwiIb6BSYQGJvqjSbHuOuhuAF9pfA1p6SgF6Bvv1uIYVgyLAgnlSFJEvRgAgf8F3FKH4AQH8FeeZpNh6D9C+enk6Oea+cwZCHwl2aYZ1o3vnu5j1o1qvoHtoSOcdshbfGBfSBCjAFbLp3SC1hZA4ZtvLUvpkZ1qiHfoQCLoYB3sBa7CRZgBRbReGAxZ0bfAKEYCzt2gehEYSr4j/l9pJbMufrwgpd9upZfNaO4gAFIGA7B86IZeWeWQnOQrnbly5/4QQAlI6iqi4sqJ6Hxa6uEKR5J0gF0nFqRYVgcGljgEDTBrpUWhXsTQxVXroQQNWsqAWQnKXBW2y6WgJdbcn7hKWB6Db4XvXymlGf6YBvWyWpwJ4dzxtNb8QBg7X1XNWa6uZnWlHdmgpWpA3oYg3fQQ2bjR7WXoXU6Er87KXuX3bZWgZKWqbHWy3i7ZW+WwxXWrX3WYAtZwZa2kxw2lGJ4kR8rNWMg+VIsrkpilD1KaXWiU363fXJHwQRHP79JpWQhs3ZwQ2w3vXJVAGHWo6enB29VTjR3m30X8112Qhp3QXpHr6IXM2CLU2BXD3aXW36XZHZGynyNfQRHn3BK33kw0wz2P6rKb35nDwlmXpHbSGG91mJgrQkgoa4qErYPcB9m3xYXjnuYkOiWybgxsHQOZ3Nm+xU7wPyGEOkgUOEn/XJH77DnUOFY25eKcPQXwQ4BfYlkznwGFDWP8L4L7ws4GL4qpQOrcBMOG3KgBPeJ4IQraapxdUoDXldbZPCWBOzbsRLbXln2O2u2pbjwnpi3N3AR3nsMROtP+A3pjGKmBbv2uQ0xxgWAs42BAMOJYA6sIB8OIJ/ceg7OgNscyZdokhROCymaY491ISZ9skogRFBqlLyNbKBVuOrRTLy265qqJaAOyxoumhYuUJ6K9yMApRqQnApWZXW744bO4uPOHOZ7JgXPqQ3OMAyuvOkvoPdOdB9OdFKg8vWRIub20uqqePsvcvVJSNdy/D+unAgmLmMLivbP7PscnOquFhhZavpvYBvPGvFoWueg2uBvbQmuMB1uMBNvRvgmO7cJuuMvAQsvhuCIBvUZLu8u3GGKrunBSAhOjJJvSuluKvnPOZXOFu6vluGuUJKGrQ9uDvWQgeUIQe7vtu1vVIPn9u8vnvOv4L4ZzpEZLoUYhvbo9KMA6ILucp4q2CkJ0OoGXvyWwwoHnbGBCBCAidQY05NHzGmISWu36a2YOYCO+YBYhYRYObsgubpZZYSfTO3SWKuxTGZnGf6WGBvCt6vYu3YZECUfrwLpWqbvbpKhxe3hkanoi3fOjOi3hfKnLPGale3gVebY+u+kcY2naYrQSYkgbfRaUIkewiioLALYzple0fVfMe7YHY6RTI8e+ACeBPSfQnoaOqVIaQ3J4JFfLZUfrYMeGK8gHpFPahlOzXXGQ4wJpPnpcO6BXSKeOXNfOqfFtGgJXTn2SfJPOGXQo+90dfuYdO4Z4+vfE/Bvk+NeGefFG/3o9ekJe+EATOv2b3NPrx6+XJU+6/AuXIlOLazXAnm+4/PezfveLfhurfcZ8Z2niZSYGBHeOmrRheaxI+Z/OPcIgOOyYeWbWuHuOqw+82Oqqa5SuK5foY8z/KXQR/Lnl5cByhiMsq4wIqng1ubYZN4zAd7NUV4D+4mg/KDdA0hYhdgn+bycYMz3f5pgOq3/L2BPB8KE83ggBUATonAHkBIBS1QnJbggAyp7KSCQkMV0QFhhkBxAttIaQ+RoDeW0tLhvLVNCK1lar0VuNDHeIKFpiNTFhIPgErQx/gQUY1kHzfADwXU7aCuEOHU7esgYPEEdsCXrYacVQYTVRgbWxhJAsYwvBdt6x0G8N5KYnamk0CWR6MsyuYZkGzU4B0QjBKodQeLUgriDeWy7RLnU30YNNDeFneApgJS7UA/QQAA===))
```asm
LoggerMessage+<>c__DisplayClass0_0.<Define>b__1(ILogger, System.Exception)
    L0000: jmp LoggerMessage+<>c__DisplayClass0_0.<Define>g__Log|0(ILogger, System.Exception)

LoggerMessage+<>c__DisplayClass0_0.<Define>b__2(ILogger, System.Exception)
    L0000: push rdi
    L0001: push rsi
    L0002: push rbx
    L0003: sub rsp, 0x20
    L0007: mov rsi, rcx
    L000a: mov rdi, rdx
    L000d: mov rbx, r8
    L0010: mov edx, [rsi+0x10]
    L0013: mov rcx, rdi
    L0016: mov r11, 0x7ff7a9a07020
    L0020: mov rax, 0x7ff7a9a07020
    L002a: call qword ptr [rax]
    L002c: test eax, eax
    L002e: je short L0045
    L0030: mov rcx, rsi
    L0033: mov rdx, rdi
    L0036: mov r8, rbx
    L0039: add rsp, 0x20
    L003d: pop rbx
    L003e: pop rsi
    L003f: pop rdi
    L0040: jmp LoggerMessage+<>c__DisplayClass0_0.<Define>g__Log|0(ILogger, System.Exception)
    L0045: add rsp, 0x20
    L0049: pop rbx
    L004a: pop rsi
    L004b: pop rdi
    L004c: ret
```

AggressiveInlining doesn't change anything here.

Comment (davidfowl) [reply src/libraries/Microsoft.Extensions.Logging.Abstractions/src/LoggerMessage.cs:159]: Grrr. Can you manually inline then?

Comment (davidfowl) [reply src/libraries/Microsoft.Extensions.Logging.Abstractions/src/LoggerMessage.cs:159]: Does it change if you do this:

```C#
if (skipEnabledCheck)
{
    return Log;
}
```

**Review by maryamariyan — APPROVED** [review]

File: src/libraries/Microsoft.Extensions.Logging.Abstractions/src/LoggerMessage.cs (hunk 1 above)
Comment (davidfowl) [reply src/libraries/Microsoft.Extensions.Logging.Abstractions/src/LoggerMessage.cs:159]: Lets change this to return the Log method directly for the skipEnabledCheck

Comment (gfoidl) [reply src/libraries/Microsoft.Extensions.Logging.Abstractions/src/LoggerMessage.cs:159]: Done with 3e547f15c3a976f70c166ac36c7ad8b70d470469

The other branch (!skipEnabledCheck) is left as is. Codegen (from above) is the check for `IsEnabled` then the tail-call. I assume together with PGO machine-code layout will be good, so there should be no perf-penalty.
On local micro-benchmarks (with a Null-Logger) I couldn't see any difference between calling `Log(...)` local function or manually inlining.

With the suggested change to return `Log` directly for the skipEnabledCheck the IL size is smaller too and the indirection for the tail-call is gone.

**Comment by davidfowl** [comment]
thanks @gfoidl, stellar work as usual!

**Comment by maryamariyan** [comment]
@gfoidl we forgot to add tests for the new APIs. Would you be open to adding a PR for that?

**Comment by gfoidl** [comment]
Sure, will address this over the weekend.

**Comment by stephentoub** [comment]
What is the cost of the check this is skipping? Is it not just a compare and a branch? Are there benchmarks that show these overloads are worth it?

**Comment by davidfowl** [comment]
Delegate invocations can't be inlined. So the cost is a delegate invocation unless an explicit check is added on the outside. This just removes the redundant check when IsEnabled is being used

**Comment by davidfowl** [comment]
Also FWIW, the logging system needs to ask multiple providers if IsEnabled is true. It isn't a simple check. First there's a filtering at the logger factory level and then it needs to call into each provider to ask it if it's enabled. That's up to n interface dispatches per call. Repeating it is just redundant when you know you're going to check it before the call.

**Comment by stephentoub** [comment]
Thanks.  Is there a typical example as would show up in a real app that highlights these costs well? Do we have benchmarks around this somewhere?

**Comment by davidfowl** [comment]
https://github.com/dotnet/runtime/issues/48619

File: src/libraries/Microsoft.Extensions.Logging.Abstractions/src/LoggerMessage.cs (hunk 1 above)
Comment (maryamariyan) [reply src/libraries/Microsoft.Extensions.Logging.Abstractions/src/LoggerMessage.cs:159]: @gfoidl did you find micro benchmark use cases for which adding a skipEnabledCheck would have improved performance?

**Pull request merged**